evm/src/recursive_verifier.rs

```diff
@@ -1,148 +1,574 @@
+use crate::cross_table_lookup::{
+    verify_cross_table_lookups, verify_cross_table_lookups_circuit, CrossTableLookup,
+    CtlCheckVarsTarget,
+};
 use crate::keccak::keccak_stark::KeccakStark;
 use crate::keccak_memory::keccak_memory_stark::KeccakMemoryStark;
 use crate::logic::LogicStark;
 use crate::memory::memory_stark::MemoryStark;
-use crate::permutation::PermutationCheckDataTarget;
+use crate::permutation::{
+    get_grand_product_challenge_set, get_grand_product_challenge_set_target, GrandProductChallenge,
+    GrandProductChallengeSet, PermutationCheckDataTarget,
+};
 use crate::proof::{
-    AllProof, AllProofChallengesTarget, AllProofTarget, BlockMetadata, BlockMetadataTarget,
-    PublicValues, PublicValuesTarget, StarkOpeningSetTarget, StarkProof,
-    StarkProofChallengesTarget, StarkProofTarget, TrieRoots, TrieRootsTarget,
+    AllChallengerState, AllProof, AllProofTarget, BlockMetadata, BlockMetadataTarget, PublicValues,
+    PublicValuesTarget, StarkOpeningSetTarget, StarkProof, StarkProofChallengesTarget,
+    StarkProofTarget, TrieRoots, TrieRootsTarget,
 };
 use crate::stark::Stark;
 use crate::util::{h160_limbs, u256_limbs};
 use crate::vanishing_poly::eval_vanishing_poly_circuit;
 use crate::vars::StarkEvaluationTargets;
 
-pub fn verify_proof_circuit<
+/// Table-wise recursive proofs of an `AllProof`.
+pub struct RecursiveAllProof<
+    F: RichField + Extendable<D>,
+    C: GenericConfig<D, F = F>,
+    const D: usize,
+> {
+    pub recursive_proofs: [ProofWithPublicInputs<F, C, D>; NUM_TABLES],
+    pub cross_table_lookups: Vec<CrossTableLookup<F>>,
+}
+
+struct PublicInputs<T: Copy + Eq + PartialEq + Debug> {
+    trace_cap: Vec<Vec<T>>,
+    ctl_zs_last: Vec<T>,
+    ctl_challenges: GrandProductChallengeSet<T>,
+    challenger_state_before: [T; SPONGE_WIDTH],
+    challenger_state_after: [T; SPONGE_WIDTH],
+}
+
+impl<T: Copy + Eq + PartialEq + Debug> PublicInputs<T> {
+    fn from_vec(v: &[T], config: &StarkConfig) -> Self {
+        let mut start = 0;
+        let trace_cap = v[start..4 * (1 << config.fri_config.cap_height)]
+            .chunks(4)
+            .map(|chunk| chunk.to_vec())
+            .collect();
+        start += 4 * (1 << config.fri_config.cap_height);
+        let ctl_challenges = GrandProductChallengeSet {
+            challenges: (0..config.num_challenges)
+                .map(|i| GrandProductChallenge {
+                    beta: v[start + 2 * i],
+                    gamma: v[start + 2 * i + 1],
+                })
+                .collect(),
+        };
+        start += 2 * config.num_challenges;
+        let challenger_state_before = v[start..start + SPONGE_WIDTH].try_into().unwrap();
+        let challenger_state_after = v[start + SPONGE_WIDTH..start + 2 * SPONGE_WIDTH]
+            .try_into()
+            .unwrap();
+
+        start += 2 * SPONGE_WIDTH;
+        let ctl_zs_last = v[start..].to_vec();
```

Comment:
I think this would be nicer with an iterator, although it'd be easier if `next_chunk` was stable, but we could use a helper for now -
```suggestion
/// Similar to the unstable `Iterator::next_chunk`. Could be replaced with that when it's stable.
fn next_chunk<T: Debug, const N: usize>(iter: &mut impl Iterator<Item = T>) -> [T; N] {
    (0..N)
        .flat_map(|_| iter.next())
        .collect_vec()
        .try_into()
        .expect("Not enough elements")
}

impl<T: Copy + Eq + PartialEq + Debug> PublicInputs<T> {
    fn from_vec(v: &[T], config: &StarkConfig) -> Self {
        let mut iter = v.iter().copied();
        let trace_cap = (0..1 << config.fri_config.cap_height)
            .map(|_| next_chunk::<_, 4>(&mut iter).to_vec())
            .collect();
        let ctl_challenges = GrandProductChallengeSet {
            challenges: (0..config.num_challenges)
                .map(|_| GrandProductChallenge {
                    beta: iter.next().unwrap(),
                    gamma: iter.next().unwrap(),
                })
                .collect(),
        };
        let challenger_state_before = next_chunk(&mut iter);
        let challenger_state_after = next_chunk(&mut iter);
        let ctl_zs_last = iter.collect();
```

Reply:
Very nice 👍